File: native/OrbiPac/Game/PacMan.swift
import Foundation
import CoreGraphics

final class PacMan {
    var ring, targetRing: Int; var radius: CGFloat; var facing: Dir = .up
    var facingAngle: CGFloat = -.pi/2, mouth: CGFloat = -.pi/2, alive = true
    private var moving=false, alignRot: CGFloat?, pendingCross: (Int, Dir)?
    private var radialCool: CGFloat = 0
    /// Continuous outer-ring cruise in one direction (radians of maze rotation).
    private var outerLoopProgress: CGFloat = 0
    private var outerLoopSign = 0
    private var mirrorLoopReady = false
    private let faceAngles: [Dir:CGFloat] = [.right:0,.down:.pi/2,.left:.pi,.up:-.pi/2]
    init(_ maze: Maze) { ring=maze.ringCount-1; targetRing=ring; radius=maze.ringRadius(ring) }
    func reset(_ maze: Maze) {
        ring=maze.ringCount-1; targetRing=ring; radius=maze.ringRadius(ring)
        facing = .up; facingAngle = -.pi/2; mouth = -.pi/2; alive=true
        moving=false; alignRot=nil; pendingCross=nil; radialCool=0
        resetOuterLoop()
    }
    var isMoving: Bool { moving || alignRot != nil }
    func localAngle(_ maze: Maze)->CGFloat { maze.toLocal(Constants.pacScreenAngle) }
    func occupancyRing(_ maze: Maze)->Int { guard moving else{return ring}; return abs(radius-maze.ringRadius(ring)) < abs(radius-maze.ringRadius(targetRing)) ? ring : targetRing }

    /// True once after a full 360° cruise on the outermost ring in one continuous move.
    func consumeMirrorLoopTrigger() -> Bool {
        defer { mirrorLoopReady = false }
        return mirrorLoopReady
    }

    func resetOuterLoop() {
        outerLoopProgress = 0
        outerLoopSign = 0
        mirrorLoopReady = false
    }

    func update(_ maze: Maze,_ dt: CGFloat,_ held:Set<Dir>,_ impulse:CGFloat) {
        guard alive else{return}; var active=false
        if let align=alignRot, let pending=pendingCross {
            resetOuterLoop()
            let d=Constants.delta(maze.rotation,align), step=CGFloat(4.2)*dt
            if abs(d) <= step { maze.rotation=align; alignRot=nil; targetRing=pending.0; facing=pending.1; pendingCross=nil; moving=true } else { maze.rotate((d < 0 ? -1:1)*step) }; active=true
        } else if !moving {
            var rot=impulse; let omega=Constants.pacLinearSpeed/max(24,radius)
            if held.contains(.left) {rot -= omega*dt}; if held.contains(.right) {rot += omega*dt}
            var applied: CGFloat = 0
            if rot != 0 {
                rot=maze.clampRotationForPac(ring,localAngle(maze),rot)
                if rot != 0 {
                    maze.rotate(rot)
                    applied = rot
                    facing=rot < 0 ? .left:.right
                    active=true
                }
            }
            trackOuterLoop(maze: maze, held: held, applied: applied)

            let local=localAngle(maze)
            radialCool = max(0, radialCool - dt)
            if radialCool <= 0 {
                if held.contains(.up),maze.canPacCross(ring,true,local) {
                    resetOuterLoop()
                    let mid=maze.nearestGapMid(maze.wallIndexForPacCross(ring,true),local)
                    alignRot=Constants.norm(Constants.pacScreenAngle-mid)
                    pendingCross=(ring-1,.up); facing = .up; active=true
                } else if held.contains(.down),maze.canPacCross(ring,false,local) {
                    resetOuterLoop()
                    let mid=maze.nearestGapMid(maze.wallIndexForPacCross(ring,false),local)
                    alignRot=Constants.norm(Constants.pacScreenAngle-mid)
                    pendingCross=(ring+1,.down); facing = .down; active=true
                }
            }
        }
        if moving {
            resetOuterLoop()
            let dest=maze.ringRadius(targetRing), step=Constants.pacLinearSpeed*dt, delta=dest-radius
            if abs(delta) <= step {radius=dest;ring=targetRing;moving=false;radialCool=Constants.radialCrossCooldown} else {radius += (delta < 0 ? -step:step)}
            facing=delta < 0 ? .up:.down; active=true
        }
        let target=faceAngles[facing]!; let d=Constants.delta(facingAngle,target), step=14*dt
        facingAngle = abs(d) <= step ? target : Constants.norm(facingAngle + (d < 0 ? -step:step))
        mouth = active ? mouth + dt*22 : -.pi/2
    }

    private func trackOuterLoop(maze: Maze, held: Set<Dir>, applied: CGFloat) {
        let outer = maze.ringCount - 1
        let holdingTurn = held.contains(.left) || held.contains(.right)
        guard ring == outer, holdingTurn else {
            resetOuterLoop()
            return
        }
        guard applied != 0 else { return } // blocked by spoke — stall, don't reset
        let sign = applied < 0 ? -1 : 1
        if outerLoopSign != 0 && sign != outerLoopSign {
            outerLoopProgress = 0
        }
        outerLoopSign = sign
        outerLoopProgress += abs(applied)
        if outerLoopProgress >= .pi * 2 {
            mirrorLoopReady = true
            outerLoopProgress = 0
            outerLoopSign = 0
        }
    }
}

File: src/game/PacMan.ts
import {
  PAC_LINEAR_SPEED,
  PAC_SCREEN_ANGLE,
  RADIAL_CROSS_COOLDOWN,
  angleDelta,
  normAngle,
} from './constants'
import type { Maze } from './Maze'
import type { Dir } from './types'

const FACE_ANGLES: Record<Dir, number> = {
  right: 0,
  down: Math.PI / 2,
  left: Math.PI,
  up: -Math.PI / 2,
}

/** How quickly Pac's mouth turns toward a new heading (rad/sec). */
const FACE_TURN_SPEED = 14
/** How quickly the maze rotates to center a gap before a radial cross. */
const ALIGN_SPEED = 4.2

export class PacMan {
  ring: number
  targetRing: number
  radius: number
  /** Logical facing used by ghost AI. */
  facing: Dir = 'up'
  /** Smoothed mouth angle for rendering. */
  facingAngle = FACE_ANGLES.up
  /** Chomp phase; −π/2 = fully closed. */
  mouth = -Math.PI / 2
  alive = true
  private moving = false
  /** Maze rotation we're easing toward before a ring cross. */
  private alignRot: number | null = null
  private pendingCross: { targetRing: number; facing: Dir } | null = null
  /** Seconds left before another up/down ring cross can start. */
  private radialCool = 0

  constructor(maze: Maze) {
    this.ring = maze.ringCount - 1
    this.targetRing = this.ring
    this.radius = maze.ringRadius(this.ring)
  }

  reset(maze: Maze): void {
    this.ring = maze.ringCount - 1
    this.targetRing = this.ring
    this.radius = maze.ringRadius(this.ring)
    this.facing = 'up'
    this.facingAngle = FACE_ANGLES.up
    this.mouth = -Math.PI / 2 // closed
    this.alive = true
    this.moving = false
    this.alignRot = null
    this.pendingCross = null
    this.radialCool = 0
  }

  get isMoving(): boolean {
    return this.moving || this.alignRot !== null
  }

  localAngle(maze: Maze): number {
    return maze.toLocal(PAC_SCREEN_ANGLE)
  }

  occupancyRing(maze: Maze): number {
    if (!this.moving) return this.ring
    const a = maze.ringRadius(this.ring)
    const b = maze.ringRadius(this.targetRing)
    return Math.abs(this.radius - a) < Math.abs(this.radius - b) ? this.ring : this.targetRing
  }

  private setFacing(dir: Dir): void {
    this.facing = dir
  }

  private easeFacing(dt: number): void {
    const target = FACE_ANGLES[this.facing]
    const d = angleDelta(this.facingAngle, target)
    const step = FACE_TURN_SPEED * dt
    if (Math.abs(d) <= step) this.facingAngle = target
    else this.facingAngle = normAngle(this.facingAngle + Math.sign(d) * step)
  }

  /** Angular maze turn that yields constant linear pace past Pac. */
  private rotateOmega(): number {
    return PAC_LINEAR_SPEED / Math.max(24, this.radius)
  }

  update(maze: Maze, dt: number, held: Set<Dir>, rotateImpulse: number): void {
    if (!this.alive) return

    let active = false

    if (this.alignRot !== null && this.pendingCross) {
      const d = angleDelta(maze.rotation, this.alignRot)
      const step = ALIGN_SPEED * dt
      if (Math.abs(d) <= step) {
        maze.rotation = this.alignRot
        this.alignRot = null
        this.targetRing = this.pendingCross.targetRing
        this.setFacing(this.pendingCross.facing)
        this.pendingCross = null
        this.moving = true
      } else {
        maze.rotate(Math.sign(d) * step)
      }
      active = true
    } else if (!this.moving) {
      const omega = this.rotateOmega()
      let rot = rotateImpulse
      if (held.has('left')) rot -= omega * dt
      if (held.has('right')) rot += omega * dt

      if (rot !== 0) {
        const local = this.localAngle(maze)
        rot = maze.clampRotationForPac(this.ring, local, rot)
        if (rot !== 0) {
          maze.rotate(rot)
          this.setFacing(rot < 0 ? 'left' : 'right')
          active = true
        }
      }

      const local = this.localAngle(maze)
      this.radialCool = Math.max(0, this.radialCool - dt)

      if (this.radialCool <= 0) {
        if (held.has('up') && maze.canPacCross(this.ring, true, local)) {
          const wall = maze.wallIndexForPacCross(this.ring, true)
          const mid = maze.nearestGapMid(wall, local)
          this.alignRot = normAngle(PAC_SCREEN_ANGLE - mid)
          this.pendingCross = { targetRing: this.ring - 1, facing: 'up' }
          this.setFacing('up')
          active = true
        } else if (held.has('down') && maze.canPacCross(this.ring, false, local)) {
          const wall = maze.wallIndexForPacCross(this.ring, false)
          const mid = maze.nearestGapMid(wall, local)
          this.alignRot = normAngle(PAC_SCREEN_ANGLE - mid)
          this.pendingCross = { targetRing: this.ring + 1, facing: 'down' }
          this.setFacing('down')
          active = true
        }
      }
    }

    if (this.moving) {
      const dest = maze.ringRadius(this.targetRing)
      const step = PAC_LINEAR_SPEED * dt
      const delta = dest - this.radius
      if (Math.abs(delta) <= step) {
        this.radius = dest
        this.ring = this.targetRing
        this.moving = false
        this.radialCool = RADIAL_CROSS_COOLDOWN
      } else {
        this.radius += Math.sign(delta) * step
      }
      // Face the radial travel direction while crossing
      this.setFacing(delta < 0 ? 'up' : 'down')
      active = true
    }

    // Mouth aims with movement; chomp only while actively moving
    this.easeFacing(dt)
    if (active) this.mouth += dt * 22
    else this.mouth = -Math.PI / 2
  }
}
